test/Test/Fission/Mock/Types.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Test.Fission.Mock.Types
+  ( Mock (..)
+  , MockSession (..)
+  , RunDB (..)
+  , GetVerifier (..)
+  ) where
+
+import           Control.Monad.Writer
+import           Data.Generics.Product
+import Database.Esqueleto
+import qualified Network.IPFS.Types  as IPFS
+import           Servant
+
+import           Test.Fission.Mock.Effect
+
+import           Fission.Prelude
+import           Fission.IPFS.Linked.Class
+import           Fission.Web.Auth.Class
+import           Fission.Models
+
+import qualified Fission.Platform.Heroku.Auth.Types as Heroku
+
+-- | The result of running a mocked test session
+data MockSession log a = MockSession
+  { effectLog :: [OpenUnion log] -- ^ List of effects that were run
+  , result    :: a               -- ^ Pure return value
+  }
+
+{- | Fission's mock type
+
+     Notes:
+     * We will likely want @State@ in here at some stage
+     * @RIO@ because lots of constraints want @MonadIO@
+       * Avoid actual @IO@, or we're going to have to rework this 😉
+
+-}
+newtype Mock effs ctx a = Mock
```

Comment:
Testing pure code is much easier than all the things with side effects. This patterns turns side-effecting code into pure code.

We can then log what would have happened, and return fake values from e.g. a fake network call. So, for instance, in a complex handler that does a bunch of DB and AWS stuff: instead of actually trying to hit Route53 and DB setup/teardown it'll both return the correct response that the server would have responded with, and push `UpdateRoute53` and `CreatedUser` into the list of which effects were run. We then write tests for both the return values, and that it would have actually hit AWS and the database.

Because of how types work, we can use these mocks, and then separately test the actual DB access or whatever, and know that the test correctness composes (in a way that it just doesn't in languages with weaker type systems).

Reply:
Amazing 🤯 👏